ast.ts

```diff
@@ -7,6 +7,7 @@ export type Type =
   | {tag: "none"}
   | {tag: "class", name: string}
   | {tag: "either", left: Type, right: Type }
+  | {tag: "generator", type: Type } // generator type
```

Comment:
Your generator still has a type in the abstract syntax but not one in the concrete syntax.

Reply:
Could you please clarify what a concrete syntax means? `type` is the type of elements generated by this generator.
E.g. `(a for a in [1, 2, 3])` would be a generator with type `number`.